docs/specification.md

```diff
@@ -0,0 +1,555 @@
+  * negative in `Y`: this amount will be `transfer`red **to** `to_y`
+- If the position update is no longer acceptable because the `deadline` was not
+  met, fails with `TIMED_OUT`
+- If the amount of tokens that needs to be `transfer`red to the contract is
+  higher than `maximum_tokens_contributed`, fails with `HIGH_TOKENS`.
+
+```ocaml
+type set_position_param = {
+    i_l : tick_index ;
+    i_u : tick_index ;
+    i_l_l : tick_index ;
+    i_u_l : tick_index ;
+    delta_liquidity : int ;
+    to_x : address ;
+    to_y : address ;
+    deadline : timestamp ;
+    maximum_tokens_contributed : nat ;
+}
+```
+
+### **snapshot_cumulatives_inside**
+
+Oracle `view` for a snapshot of the tick cumulative, seconds per liquidity and
+seconds inside the given range.
+
+- `i_l` determines the lowest tick index of the range.
+- `i_u` determines the highest tick index of the range.
+- the `callback` contract will be called with the computed values.
+- `tick_cumulative_inside` is the computed snapshot of the tick cumulative for
+  the given range.
+- `seconds_per_liquidity_inside` is the computed snapshot of the seconds per
+  liquidity for the given range.
+- `seconds_inside` is the computed snapshot of the seconds for the given range.
+
+```ocaml
+type sci_view_param = {
+    tick_cumulative_inside : int ;
+    seconds_per_liquidity_inside : nat ;
+    seconds_inside : nat ;
+}
+
+type sci_param = {
+    i_l : tick_index ;
+    i_u : tick_index ;
+    callback : sci_view_param contract ;
+}
+```
+
+
+### **observe**
+
+Oracle `view` for the cumulative tick and liquidity-in-range, calculated for each
+`timestamp`s in a list.
+
+- Each value in `times` will be used to calculate a `cumulative_entry`.
+  The order of the computed entries is the same as that of the given `timestamp`s.
+- The `callback` contract will be called with the computed values.
+- For each `timestamp`/`cumulative_entry` pair:
+  * `tick_cumulative` is the cumulative tick value at that `timestamp`.
+  * `seconds_per_liquidity_cumulative` is the cumulative seconds per
+    liquidity-in-range at that `timestamp`.
+- If any of the given timestamps is too far back in the past,
+  fails with `INVALID_TIMESTAMP`.
+  Note that the amount of observations stored by the contract can be increased
+  via the `increase_observation_count` entrypoint.
+
+```ocaml
+type cumulative_entry = {
+    tick_cumulative : int ;
+    seconds_per_liquidity_cumulative : nat ;
+}
+
+type oracle_view_param = cumulative_entry list
+
+type observe_param = {
+    times : timestamp list ;
+    callback : oracle_view_param contract ;
+}
+```
+
+### **increase_observation_count**
+
+Increase the number of observations of `tick_cumulative` and
+`seconds_per_liquidity_cumulative` taken and stored in the contract.
+The greater the number, the further back in time users will be able to
+retrieve data using [`observe`][#observe].
+
+The caller of this entrypoint will pay for the additional storage costs.
```

Comment:
>  I suppose we won't just initialize `new observation_count - old observation_count`` slots with dummy values, or we will?

Haven't thought too much about it yet, but yes, that sounds like the most probable scenario.

Circular buffers are usually implemented by having a fixed-size array, and two pointers to where the actual contents begin and end (`lo` and hi`), and making sure `lo` and `hi` never cross each other. It doesn't matter what's stored in the slots above `hi` or below `lo`, because reading those values would be considered an "out of bounds access".

In our case, the fixed size array will be a big_map emulating a linked list. But, apart from that, I think the usual strategy should work the same.

Reply:
Thanks, your explanation sounds reasonable.

I guess later, if the first version of the implementation suddenly turns out to diverge from the spec here, let's consider this issue there.